メモアプリを新規追加

ItohYuto opened the pull request:
- 開発環境と本番環境を分けるように実装。
- ページのヘッダー部分はlayout.erbでまとめる。
- 存在しないURLへのアクセス時と存在しないメモを指定したときのエラーハンドリングを実装。
- メモのデータは/data/memo.csvで管理する。
- XSS対策済

ItohYuto committed:
メモアプリを新規追加。

- 開発環境と本番環境を分けるように実装。
- ページのヘッダー部分はlayout.erbでまとめる。
- 存在しないURLへのアクセス時と存在しないメモを指定したときのエラーハンドリングを実装。
- メモのデータは/data/memo.csvで管理する。
- XSS対策。

ItohYuto committed:
実行ファイルに実行権限を付与。

ItohYuto committed:
メモの追加時にnotes_not_foundページが表示される不具合を修正。

maedana commented on app.rb:
```diff
@@ -0,0 +1,123 @@
+                @notes.max_by { |note| note[:id] }[:id] + 1
+              end
+    CSV.open(data_file_path, 'a') do |csv|
+      csv << [note_id.to_i, h(params[:title]), h(params[:content])]
```
XSS対策は通常表示時にやるのがセオリーかなと思います。サーバサイドでユーザ入力値を加工してしまうと、多重エスケープされたり、あとから仕様変更で一部タグは許容したいとかあると困るなど色々と困りことが起こりがちかなと。

maedana commented on app.rb:
```diff
@@ -0,0 +1,123 @@
+        @note = note
+      end
+      next if note[:id] == note_id.to_i
+    end
```
1件探したいだけですよね。冒頭で1件を https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/detect.html 等使って探すのがわかりやすいかなとおもいますけどいかがでしょう? 編集画面も同様です。

ItohYuto replied:
ループからの脱出処理等も不要になるのでfindメソッドを使用したほうが分かりやすいですね。
修正しました。

maedana commented on app.rb:
```diff
@@ -0,0 +1,123 @@
+        @sub_title = note[:title]
+        @note = note
+      end
+      next if note[:id] == note_id.to_i
```
nextの意味がないような。ループ抜けるならbreakでは?

ItohYuto replied:
おっしゃる通りですね。
後ろに処理がないので`next`を使用する意味がないですね。
こちらに関しては別の修正で、ループ処理を使用しなくなったため削除しました。

maedana commented on app.rb:
```diff
@@ -0,0 +1,123 @@
+    CSV.open(data_file_path, 'w', headers: true) do |csv|
+      csv << CSV_HEADER
+      deleted_notes.each { |note| csv << note }
+    end
```
書き込み部分は更新と同じですかね? もしそうなら共通化したいかなーと思いました。

ItohYuto replied:
ヘルパーに追加し、共通化しました。

maedana commented on views/layout.erb:
```diff
@@ -0,0 +1,23 @@
+<body>
+  <header>
+    <h1 class="title"><%= @title %></h1>
+    <% if @sub_title == "一覧" %>
```
この判定方法だと結構壊れやすいように見えます。パスで判断とか別のもっと壊れにくい手段を使いたいかなと思いましたがいかがでしょう?

ItohYuto replied:
リクエストパスを参照するように修正しました。

maedana commented on views/layout.erb:
```diff
@@ -0,0 +1,23 @@
+  <header>
+    <h1 class="title"><%= @title %></h1>
+    <% if @sub_title == "一覧" %>
+      <button class="btn header_btn" onclick="location.href='/register'">追加</button>
```
絶対NGとは言いませんが、やってることを考えるとわざわざjs使いたくないですね。aタグにしてボタン風のスタイルを当てるのがセオリーかなと思いました。

ItohYuto replied:
見た目のためだけにbuttonタグ使用しているのでaタグに変更しました。
実際にやりたいことはページ遷移なので、aタグのが良いという認識でお間違いないでしょうか？

maedana replied:
> 実際にやりたいことはページ遷移なので、aタグのが良いという認識でお間違いないでしょうか？

はい、その認識です。せっかくそのためのタグが存在するのに見た目のためだけにボタン使うなら見た目をcssで整えつつaタグ使うほうがよいかなということですね。

maedana commented on views/layout.erb:
```diff
@@ -0,0 +1,23 @@
+    <% if @sub_title == "一覧" %>
+      <button class="btn header_btn" onclick="location.href='/register'">追加</button>
+    <% else %>
+      <button class="btn header_btn" onclick="location.href='/'">一覧へ戻る</button>
```
ここも

ItohYuto replied:
修正しました。

maedana commented on views/view.erb:
```diff
@@ -0,0 +1,10 @@
+  <p><%= @note[:content] %></p>
+</div>
+<div class="btn-section">
+  <button class="btn" onclick="location.href='/notes/<%= @note[:id] %>/edit'">編集</button>
```
同上です。このためだけのjsは避けたいところかなーと思いました。

ItohYuto replied:
修正しました。

ItohYuto committed:
指摘箇所修正

- XSS対策の実装箇所修正。
  データ格納時からHTML出力前に変更。
- 対象の`note`を検索する処理の修正。
  findメソッドを使用いてループ処理を削除。
- csvへの書き込み処理を共通化。
  ヘルパーに`writing_into_csv`を作成。
- buttonタグでページ遷移している箇所をaタグに変更。
  cssでボタン風に変更。
- サブタイトルで判定している個所を、リクエストパスで判定するように修正。

ItohYuto committed:
erblintの警告修正

ItohYuto commented:
修正しました。
再度レビューお願いいたします。

## 各ページのスクリーンショット
- トップページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/667a1c81-bf6b-4751-bf9e-85fbd3080a43)
- メモ詳細ページ 
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/fc4981a4-fe2f-48e0-bcac-7ea8d268228f)
- メモ編集ページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/a8ee237e-406a-4eb5-90fd-adc8fda505ec)
- メモ追加ページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/fbdc63d3-083a-43e7-b6b2-0027de7cf509)
- ページが見つからなかったときのページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/cd4617ec-e747-4ed8-bb2d-3c7d1223cd7d)
- 存在しないメモにアクセスしようとしたときのページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/a171b4e8-8a51-47b5-b11f-614c8d72355d)

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/487a7e7e-a301-4846-af9b-471f7784e339)
## erb_lintの実行結果
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/504cb7a5-0d28-4ba0-96d5-27f7fb3f5957)

ItohYuto commented:
今回の修正時に編集画面及び登録画面のフォーム下部のボタンの位置を中央に配置にしております。

maedana commented on views/edit.erb:
```diff
@@ -0,0 +1,15 @@
+<form edit method="post" action="/notes/<%= @note[:id] %>">
+  <div>
+    <label for="title">タイトル</label>
+    <input type="text" id="title" name="title" value="<%= @note[:title] %>" required="true">
```
手元で動かして確認したわけじゃないのですけど、`"><script>alert("xss");</script>`みたいにしてタグを終わらせるような感じの入力をするとxss出ないですかね?

ItohYuto replied:
修正漏れでしたので、修正いたしました。

ItohYuto committed:
xss対策漏れを修正。

ItohYuto commented:
修正いたしましたので、再度レビューお願いいたします。
## 各ページのスクリーンショット
- トップページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/e4abface-dbf0-4816-89bd-976043b1f9db)
- メモ詳細ページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/2e68dea1-5905-4940-9015-b9bec4f53f2f)
- 編集ページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/972d1d53-db34-4a84-8e99-fc9a23ea813b)
- 追加ページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/e9feb50a-afd9-4e46-a136-443b39d7ddf6)
- ページが見つからなかったときのページ 
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/0c7b3d7e-2dc6-4178-8cbc-ba8a609f6771)
- 存在しないメモにアクセスしようとしたときのページ
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/e9f1fa66-9244-43f3-a770-ea03786f20c1)

## rubocop-fjordの実行結果
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/fa948a6d-571f-412b-9370-47633df952fd)

## erb_lintの実行結果
![image](https://github.com/ItohYuto/note-taking-application/assets/88817643/194830b9-eb13-48c0-977e-4e5019064f0f)

maedana reviewed (approved):
LGTM
